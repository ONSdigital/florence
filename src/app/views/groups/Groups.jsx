import React, { useEffect, useState, useCallback } from "react";
import filter from "lodash/filter";
import PropTypes from "prop-types";
import { Link } from "react-router";
import { useInput } from "../../hooks/useInput";
import url from "../../utilities/url";
import Input from "../../components/Input";
import SimpleSelectableList from "../../components/simple-selectable-list/SimpleSelectableList";
import BackButton from "../../components/back-button";

const Groups = props => {
    const { groups, isLoading, loadTeams, isNewSignIn } = props;
    const [search, setSearch] = useInput("");

    useEffect(() => {
        loadTeams(isNewSignIn);
    }, []);

    const getFilteredGroups = useCallback(() => {
        return filter(groups, group => group.name?.toLowerCase().includes(search.value.toLowerCase()));
    }, [search.value]);

    return (
        <div className="grid grid--justify-space-around">
            <div className="grid__col-11 grid__col-md-9">
                <span className="margin-top--1">
                    <h1 className="inline-block margin-top--0 margin-bottom--0 padding-right--1">Preview teams</h1>
                    <Link className="margin-left--1" to={url.resolve("./groups/create")}>
                        Create a new team
                    </Link>
                </span>
                <div className="grid__col-6">
                    <Input id="search-content-types" placeholder="Search teams by name" {...search} />
                </div>
                <SimpleSelectableList rows={search.value ? getFilteredGroups() : groups} showLoadingState={isLoading} />
            </div>
        </div>
    );
};

Groups.propTypes = {
    groups: PropTypes.array.isRequired,
    loadTeams: PropTypes.func.isRequired,
    isNewSignIn: PropTypes.bool.isRequired,
    isLoading: PropTypes.bool.isRequired,
};
export default Groups;
