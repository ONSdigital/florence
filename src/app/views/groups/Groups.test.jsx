import React from "react";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";
import { render, screen } from "../../utilities/tests/test-utils";
import Groups from "./Groups";
import "@testing-library/jest-dom/extend-expect";
import { mappedSortedGroups } from "../../utilities/tests/mockData";

const defaultProps = {
    groups: [],
    loadTeams: jest.fn(),
    isNewSignIn: true,
    isLoading: false,
};

describe("Groups", () => {
    it("matches the snapshot with empty props", () => {
        const wrapper = renderer.create(<Groups {...defaultProps} />);
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it("matches the snapshot with groups props", () => {
        const props = { ...defaultProps, groups: mappedSortedGroups };
        const wrapper = renderer.create(<Groups {...props} />);
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it("requests all teams on load", () => {
        render(<Groups {...defaultProps} />);
        expect(defaultProps.loadTeams).toHaveBeenCalledWith(true);
    });

    it("shows, message if no teams found", () => {
        render(<Groups {...defaultProps} />);
        expect(screen.getByText(/Nothing to show/i)).toBeInTheDocument();
    });

    it("shows, create button and search input", () => {
        render(<Groups {...defaultProps} />);
        expect(screen.getByText("Create a new team")).toBeInTheDocument();
        expect(screen.getByPlaceholderText("Search teams by name")).toHaveValue("");
    });

    it("shows list of teams details", () => {
        const props = { ...defaultProps, groups: mappedSortedGroups };
        render(<Groups {...props} />);

        expect(screen.getByText(/Preview teams/i)).toBeInTheDocument();
        expect(screen.getByText(/Hello Group/i)).toBeInTheDocument();
        expect(screen.getByText(/Test/i)).toBeInTheDocument();
        expect(screen.getByText(/This team has 5 members/i)).toBeInTheDocument();
        expect(screen.getByText(/Admins/i)).toBeInTheDocument();
    });

    it("filters list of teams by search term", () => {
        const getFilteredGroups = jest.fn();
        const props = { ...defaultProps, groups: mappedSortedGroups };

        render(<Groups {...props} />);

        expect(screen.getByPlaceholderText("Search teams by name")).toHaveValue("");

        userEvent.paste(screen.getByPlaceholderText(/Search teams by name/i), "Admin");
        expect(screen.getByPlaceholderText("Search teams by name")).toHaveValue("Admin");

        expect(screen.getByText(/Admins/i)).toBeInTheDocument();

        expect(screen.queryByText(/Hello Group/i)).not.toBeInTheDocument();
        expect(screen.queryByText(/Test/i)).not.toBeInTheDocument();
        expect(screen.queryByText(/This team has 5 members/i)).not.toBeInTheDocument();
    });
});
