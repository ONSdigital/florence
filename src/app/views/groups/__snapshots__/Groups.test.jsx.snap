// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Groups matches the snapshot with empty props 1`] = `
<div
  className="grid grid--justify-space-around"
>
  <div
    className="grid__col-11 grid__col-md-9"
  >
    <span
      className="margin-top--1"
    >
      <h1
        className="inline-block margin-top--0 margin-bottom--0 padding-right--1"
      >
        Preview teams
      </h1>
      <a
        className="margin-left--1"
        onClick={[Function]}
        style={Object {}}
      >
        Create a new team
      </a>
    </span>
    <div
      className="grid__col-6"
    >
      <div
        className="form__input"
      >
        <label
          className="form__label"
          htmlFor="search-content-types"
        />
        <input
          autoComplete="new-password"
          autoFocus={false}
          className="input"
          disabled={false}
          id="search-content-types"
          name="search-content-types"
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Search teams by name"
          type="text"
          value=""
        />
      </div>
    </div>
    <ul
      className="list list--neutral simple-select-list"
    >
      <p
        className="padding-top--1"
      >
        Nothing to show.
      </p>
    </ul>
  </div>
</div>
`;

exports[`Groups matches the snapshot with groups props 1`] = `
<div
  className="grid grid--justify-space-around"
>
  <div
    className="grid__col-11 grid__col-md-9"
  >
    <span
      className="margin-top--1"
    >
      <h1
        className="inline-block margin-top--0 margin-bottom--0 padding-right--1"
      >
        Preview teams
      </h1>
      <a
        className="margin-left--1"
        onClick={[Function]}
        style={Object {}}
      >
        Create a new team
      </a>
    </span>
    <div
      className="grid__col-6"
    >
      <div
        className="form__input"
      >
        <label
          className="form__label"
          htmlFor="search-content-types"
        />
        <input
          autoComplete="new-password"
          autoFocus={false}
          className="input"
          disabled={false}
          id="search-content-types"
          name="search-content-types"
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="Search teams by name"
          type="text"
          value=""
        />
      </div>
    </div>
    <ul
      className="list list--neutral simple-select-list"
    >
      <li
        className="simple-select-list__item"
      >
        <div
          className="simple-select-list__col simple-select-list__cols-1"
        >
          <a
            className="simple-select-list__title"
            onClick={[Function]}
            role="link"
            style={Object {}}
          >
            Hello Group
          </a>
        </div>
      </li>
      <li
        className="simple-select-list__item"
      >
        <div
          className="simple-select-list__col simple-select-list__cols-1"
        >
          <a
            className="simple-select-list__title"
            onClick={[Function]}
            role="link"
            style={Object {}}
          >
            Test
          </a>
        </div>
      </li>
      <li
        className="simple-select-list__item"
      >
        <div
          className="simple-select-list__col simple-select-list__cols-1"
        >
          <a
            className="simple-select-list__title"
            onClick={[Function]}
            role="link"
            style={Object {}}
          >
            This team has 5 members
          </a>
        </div>
      </li>
      <li
        className="simple-select-list__item"
      >
        <div
          className="simple-select-list__col simple-select-list__cols-1"
        >
          <a
            className="simple-select-list__title"
            onClick={[Function]}
            role="link"
            style={Object {}}
          >
            Admins
          </a>
        </div>
      </li>
    </ul>
  </div>
</div>
`;
